modules/rgbd/src/hash_tsdf.cpp

```diff
@@ -0,0 +1,712 @@
+        const Matx33f cam2volRot   = cam2vol.rotation();
+        const Matx33f vol2camRot   = vol2cam.rotation();
+
+        const float blockSize = volume.volumeUnitSize;
+
+        for (int y = range.start; y < range.end; y++)
+        {
+            ptype* ptsRow = points[y];
+            ptype* nrmRow = normals[y];
+
+            for (int x = 0; x < points.cols; x++)
+            {
+                //! Initialize default value
+                Point3f point = nan3, normal = nan3;
+
+                //! Ray origin and direction in the volume coordinate frame
+                Point3f orig    = cam2volTrans;
+                Point3f rayDirV = normalize(Vec3f(cam2volRot * reproj(Point3f(x, y, 1.f))));
+
+                float tmin = rgbd::Odometry::DEFAULT_MIN_DEPTH() / rayDirV.z;
+                float tmax = rgbd::Odometry::DEFAULT_MAX_DEPTH() / rayDirV.z;
+
+                float tprev = tmin;
+                float tcurr = tmin;
+
+                TsdfType prevTsdf = volume.truncDist;
+                cv::Vec3i prevVolumeUnitIdx =
+                    cv::Vec3i(std::numeric_limits<int>::min(), std::numeric_limits<int>::min(),
+                              std::numeric_limits<int>::min());
+                cv::Ptr<TSDFVolumeCPU> currVolumeUnit;
+                while (tcurr < tmax)
+                {
+                    Point3f currRayPos          = orig + tcurr * rayDirV;
+                    cv::Vec3i currVolumeUnitIdx = volume.volumeToVolumeUnitIdx(currRayPos);
+
+                    VolumeUnitMap::const_iterator it = volume.volumeUnits.find(currVolumeUnitIdx);
+
+                    TsdfType currTsdf = prevTsdf;
+                    int currWeight    = 0;
+                    float stepSize    = blockSize;
+                    cv::Vec3i volUnitLocalIdx;
+
+                    //! Does the subvolume exist in hashtable
+                    if (it != volume.volumeUnits.end())
+                    {
+                        currVolumeUnit =
+                            std::dynamic_pointer_cast<TSDFVolumeCPU>(it->second.pVolume);
+                        cv::Point3f currVolUnitPos =
+                            volume.volumeUnitIdxToVolume(currVolumeUnitIdx);
+                        volUnitLocalIdx = volume.volumeToVoxelCoord(currRayPos - currVolUnitPos);
+
+                        //! TODO: Figure out voxel interpolation
+                        Voxel currVoxel = currVolumeUnit->at(volUnitLocalIdx);
+                        currTsdf        = currVoxel.tsdf;
+                        /* currTsdf   = volume.interpolateVoxel(currRayPos); */
+                        /* std::cout << "Current TSDF: " << currTsdf << " interpolated: " << interpolatedTsdf << "\n"; */
+                        currWeight = currVoxel.weight;
+                        stepSize   = tstep;
+                    }
+                    //! Surface crossing
+                    if (prevTsdf > 0.f && currTsdf <= 0.f && currWeight > 0)
```

Comment:
Done.

Reply:
This is not resolved yet and there is a problem:
If surface crossing happened and `tInterp` is not a good number (which is possible if `prevTsdf` is close to `currTsdf`: this leads to Inf or even NaN if `stepSize` is also small so that `tcurr` is close to `tprev`)
OR
If the voxels are OK but their neighbours are not good so that it's impossible to get normal
THEN
We skip this step and search for the next point on a ray.
BUT we shouldn't.
A loop should be finished if we faced a surface, even if we cannot get feasible points/normals from it (in that case we should indicate it with NaNs).

Reply:
What should be done is to run loop until surface crossing, mark it with a bool, after the loop check that bool and if it's true then perform the interpolation, normal calculation and so on, if it's false then write NaNs to output pixels.

Reply:
I had fixed this by moving the break outside the (is tInterp invalid check), but it seems it was not pushed correctly for some reason. I think just breaking the loop regardless of the tInterp value at a surface crossing solves the issue.